README.md
# graphite_powershell_functions Cookbook

[![Build Status](https://travis-ci.org/tas50/chef-graphite_powershell_functions.svg?branch=master)](https://travis-ci.org/tas50/chef-graphite_powershell_functions) [![Cookbook Version](https://img.shields.io/cookbook/v/tomcat.svg)](https://supermarket.chef.io/cookbooks/graphite_powershell_functions)

A cookbook to install the Graphite PowerShell Function (<https://github.com/MattHodge/Graphite-PowerShell-Functions>) script using Chef. Sets up the script as a Windows service using the Non-Sucky Service Manager (<http://nssm.cc/>) as outlined in Matt Hodge's documentation for Graphite PowerShell Functions.

## Requirements

### Chef

- Chef 12.1+

### Platforms

- Windows

### Cookbooks

- Windows

## Attributes

- `default['graphite_powershell_functions']['CarbonServer']` FQDN of the carbon host. **Required**
- `default['graphite_powershell_functions']['CarbonServerPort']` port of the carbon host (default: `2003`).
- `default['graphite_powershell_functions']['MetricPath']` Path to prefix on the collected metrics (default: `powershell.`).
- `default['graphite_powershell_functions']['MetricSendIntervalSeconds']` Interval to grab metrics (default: 30).
- `default['graphite_powershell_functions']['TimeZoneOfGraphiteServer']` Timezone of the Graphite server (default `UTC`).
- `default['graphite_powershell_functions']['PerformanceCounters']` Array of performance counters to collect
- `default['graphite_powershell_functions']['MetricFilter']` Array of Metrics to filter
- `default['graphite_powershell_functions']['nssm_archive']` URL of the Non-Sucky Service Manager zip archive
- `default['graphite_powershell_functions']['nssm_archive_checksum']` shasum for the zip archive
- `default['graphite_powershell_functions']['install_dir']` directory to install to
- `default['graphite_powershell_functions']['service_name']` name of the service
- `default['graphite_powershell_functions']['service_description'] description of the service

## License & Authors

- Author: Tim Smith ([tsmith@chef.io](mailto:tsmith@chef.io))

```text
Copyright:: 2014-2016, Limelight Networks, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# chef-graphite_powershell_functions Cookbook CHANGELOG

This file is used to list changes made in each version of the graphite_powershell_functions cookbook.

## v2.0.0

- Require Chef 12.1 or later as Chef 11 is now EOL
- Add a new attribute `node['graphite_powershell_functions']['install_dir']` to control the install directory
- Update NSSM to 2.24
- Add new attributes `node['graphite_powershell_functions']['main_script']` and `node['graphite_powershell_functions']['main_method']` to control the execution of the script
- Add new attributes `node['graphite_powershell_functions']['service_name']` and `node['graphite_powershell_functions']['service_description']` to control the service
- Updated the license to be a SPDX compliant string in metadata.rb

## v1.2.1

- Update Travis tests to run in ChefDK
- Replace rubocop with cookstyle for preset rules
- Resolve cookstyle warnings
- Update the readme to include required chef release and cookbook depedencies and to include travis and cookbook version badges
- Add chef_version, issues_url and source_url to the metadata

## v1.2.0

- Add travis.yml for testing
- Add chefignore to limit files uploaded to the chef server
- Use standard chef .gitignore file
- Remove use of hash rockets in rubocop.yml
- Use supermarket URL in berksfile
- Add a changelog
- Add a contributing doc
- Breakout deps into groups in the Gemfile and update to the latest versions
- Dont use collectd attributes directly in the attributes since this break anyone not using collectd
- Change MetricSendIntervalSeconds to 30 seconds from 15 seconds
- Update the Graphite Powershell script
- Restart the service when the config is updated
- Add more chefspec tests

## v1.1.0

- Update to the latest upstream release of the Powershell script
- Add the bulk Graphite metric update code currently awaiting merging into the upstream Powershell script
- Reorder sanitization method in the Powershell so the hostname is not sanitized. The hostname attribute being used must be properly sanitized before being set in the config file. The default attribute implements this with a gsub of . to _
- Switch the default update time period from every 15 seconds to every 30 seconds
- Add additional tests and update the test harness
- Fix an incorrect license in the default recipe file
- Restart the service when the config file or Powershell script is updated

## v1.0.0

- Initial release


spec/default_spec.rb
require 'spec_helper'

describe 'graphite_powershell_functions::default' do
  let(:chef_run) do
    ChefSpec::ServerRunner.new(platform: 'windows', version: '2012R2') do |node|
      node.set['graphite_powershell_functions']['CarbonServer'] = 'graphite.therealtimsmith.com'
      node.set['graphite_powershell_functions']['CarbonServerPort'] = 2003
      node.set['graphite_powershell_functions']['MetricPath'] = 'servers.prod.'
    end.converge(described_recipe)
  end
  subject { chef_run }
  before do
    stub_command('C:\\windows\\system32\\WindowsPowerShell\\v1.0\\powershell.exe -NoLogo -NonInteractive -NoProfile -ExecutionPolicy RemoteSigned Get-Service -Name GraphitePowerShell').and_return(true)
  end

  it 'should create install dir' do
    expect(chef_run).to create_directory 'C:\GraphitePowershellFunctions\nssm'
  end

  it 'should template the config file' do
    expect(chef_run).to render_file('C:\GraphitePowershellFunctions\StatsToGraphiteConfig.xml')
  end

  it 'should create powershell script file' do
    expect(chef_run).to create_cookbook_file 'C:\GraphitePowershellFunctions\Graphite-Powershell.ps1'
  end

  it 'enables the service' do
    expect(chef_run).to enable_service('GraphitePowerShell')
    expect(chef_run).to start_service('GraphitePowerShell')
  end
end


attributes/default.rb
#
# Cookbook Name:: graphite_powershell_functions
# Attributes:: default
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default['graphite_powershell_functions']['CarbonServer'] = nil
default['graphite_powershell_functions']['CarbonServerPort'] = 2003
default['graphite_powershell_functions']['MetricPath'] = 'powershell.'
default['graphite_powershell_functions']['MetricSendIntervalSeconds'] = 30
default['graphite_powershell_functions']['TimeZoneOfGraphiteServer'] = 'UTC'
default['graphite_powershell_functions']['hostname'] = node.name.tr('.', '_')
default['graphite_powershell_functions']['PerformanceCounters'] = [
  'Network Interface(*)\Bytes Received/sec',
  'Network Interface(*)\Bytes Sent/sec',
  'Network Interface(*)\Packets Received Unicast/sec',
  'Network Interface(*)\Packets Sent Unicast/sec',
  'Network Interface(*)\Packets Received Non-Unicast/sec',
  'Network Interface(*)\Packets Sent Non-Unicast/sec',
  'Processor(_Total)\% Processor Time',
  'Processor(_Total)\% User Time',
  'Processor(_Total)\% Idle Time',
  'Memory\Available MBytes',
  'Memory\Pages/sec',
  'Memory\Pages Input/sec',
  'System\Processor Queue Length',
  'System\Threads',
  'System\File Write Operations/sec',
  'System\File Read Operations/sec',
  'PhysicalDisk(*)\Avg. Disk Write Queue Length',
  'PhysicalDisk(*)\Avg. Disk Read Queue Length',
  'TCPv4\Segments Received/sec',
  'TCPv4\Segments Sent/sec',
  'TCPv4\Segments Retransmitted/sec',
]
default['graphite_powershell_functions']['MetricFilter'] = [
  'isatap',
  'teredo tunneling',
]

default['graphite_powershell_functions']['nssm_archive'] = 'https://nssm.cc/release/nssm-2.24.zip'
default['graphite_powershell_functions']['nssm_archive_checksum'] = '727d1e42275c605e0f04aba98095c38a8e1e46def453cdffce42869428aa6743'

default['graphite_powershell_functions']['service_name'] = 'GraphitePowerShell'
default['graphite_powershell_functions']['service_description'] = 'Send metrics to graphite (carbon) using PowerShell'

default['graphite_powershell_functions']['install_dir'] = 'C:\GraphitePowershellFunctions'
default['graphite_powershell_functions']['main_script'] = 'Graphite-PowerShell.ps1'
default['graphite_powershell_functions']['main_method'] = 'Start-StatsToGraphite'
